File: matrixchainmultiplication.cpp
#include <iostream>
#include <climits>
using namespace std;

int matrixChainOrder(int p[], int n) {
    int m[n][n];

    // cost is 0 when multiplying one matrix
    for (int i = 1; i < n; i++)
        m[i][i] = 0;

    // L = chain length
    for (int L = 2; L < n; L++) {
        for (int i = 1; i < n - L + 1; i++) {
            int j = i + L - 1;
            m[i][j] = INT_MAX;

            for (int k = i; k < j; k++) {
                int cost = m[i][k] + m[k+1][j] 
                         + p[i-1] * p[k] * p[j];

                if (cost < m[i][j])
                    m[i][j] = cost;
            }
        }
    }

    return m[1][n-1];
}

int main() {
    int p[] = {10, 30, 5, 60};
    int n = sizeof(p)/sizeof(p[0]);

    cout << "Minimum multiplications: "
         << matrixChainOrder(p, n);

    return 0;
}

File: matrixchainmultiplication.py
import sys

def matrix_chain_order(p):
    n = len(p)
    
    # dp[i][j] = minimum cost to multiply matrices i to j
    dp = [[0 for _ in range(n)] for _ in range(n)]

    # l = chain length
    for length in range(2, n):
        for i in range(1, n - length + 1):
            j = i + length - 1
            dp[i][j] = sys.maxsize

            for k in range(i, j):
                cost = (dp[i][k] +
                        dp[k+1][j] +
                        p[i-1] * p[k] * p[j])

                dp[i][j] = min(dp[i][j], cost)

    return dp[1][n-1]


# Input
n = int(input("Enter number of dimensions: "))
p = list(map(int, input("Enter dimensions: ").split()))

print("Minimum multiplication cost:", matrix_chain_order(p))